pgml-sdks/python/pgml/pgml/collection.py

```diff
@@ -0,0 +1,771 @@
+            % (self.splitters_table)
+        )
+        run_create_or_insert_statement(conn, index_statement, autocommit=True)
+
+        index_statement = (
+            "CREATE INDEX CONCURRENTLY IF NOT EXISTS parameters_index ON %s USING GIN (parameters jsonb_path_ops);"
+            % (self.splitters_table)
+        )
+        run_create_or_insert_statement(conn, index_statement, autocommit=True)
+
+        self.pool.putconn(conn)
+
+    def _create_models_table(self) -> None:
+        """
+        This function creates a table in a PostgreSQL database with specific columns and indexes.
+        """
+        conn = self.pool.getconn()
+        self.models_table = self.name + ".models"
+        create_statement = (
+            "CREATE TABLE IF NOT EXISTS %s (\
+                            id          serial8 PRIMARY KEY, \
+                            created_at  timestamptz NOT NULL DEFAULT now(), \
+                            task        text NOT NULL, \
+                            name        text NOT NULL, \
+                            parameters  jsonb NOT NULL DEFAULT '{}'\
+                    );"
+            % (self.models_table)
+        )
+        run_create_or_insert_statement(conn, create_statement)
+
+        index_statement = (
+            "CREATE INDEX CONCURRENTLY IF NOT EXISTS created_at_index ON %s (created_at);"
+            % (self.models_table)
+        )
+        run_create_or_insert_statement(conn, index_statement, autocommit=True)
+
+        index_statement = (
+            "CREATE INDEX CONCURRENTLY IF NOT EXISTS task_index ON %s (task);"
+            % (self.models_table)
+        )
+        run_create_or_insert_statement(conn, index_statement, autocommit=True)
+
+        index_statement = (
+            "CREATE INDEX CONCURRENTLY IF NOT EXISTS name_index ON %s (name);"
+            % (self.models_table)
+        )
+        run_create_or_insert_statement(conn, index_statement, autocommit=True)
+
+        index_statement = (
+            "CREATE INDEX CONCURRENTLY IF NOT EXISTS parameters_index ON %s USING GIN (parameters jsonb_path_ops);"
+            % (self.models_table)
+        )
+        run_create_or_insert_statement(conn, index_statement, autocommit=True)
+
+        self.pool.putconn(conn)
+
+    def _create_transforms_table(self) -> None:
+        """
+        This function creates a transforms table in a PostgreSQL database and adds indexes to it.
+        """
+        conn = self.pool.getconn()
+        self.transforms_table = self.name + ".transforms"
+        create_statement = (
+            "CREATE TABLE IF NOT EXISTS %s (\
+                            oid         regclass PRIMARY KEY,\
+                            created_at  timestamptz NOT NULL DEFAULT now(), \
+                            task        text NOT NULL, \
+                            splitter    int8 NOT NULL REFERENCES %s\
+                              ON DELETE CASCADE\
```

Comment:
Are you sure you want to `CASCADE`? This has the effect of automatically deleting rows from this table when the row they are referencing in another table is deleted. This can delete a lot of data accidentally. The preferred way for me generally is to `ON DELETE RESTRICT` which is the default. That way, you'll get an error when attempting to delete a splitter that's referenced by this table. This is default behavior also, so you can just remove `ON DELETE CASCADE`.

Reply:
In these tables, input is write-once and additional rows in other tables are generated formulaically from the initial data in a repeatable way. The deletion path is invoked to remove a write-once record and everything associated with it -- it's a kind of GC, and that is basically what `CASCADE` is for.